### DISCUSSION TITLE
How to add dependency files other than ``.js`` for custom element?

### DISCUSSION THREAD

**Discussion by KiddoV** [zauberzeug/nicegui#2635] [Q&A]
### Question

I created a custom element for ``MonacoEditor``:
```txt
# folder structure
elements/
|-- monaco-editor/
|-- lib/
|---- monaco-editor.js <- main file
|---- codicon.css
|---- codicon.ttf
|-- monaco_editor.js <- Vue component
```
```py
# elements/monaco_editor.py
from typing import Dict, Optional

from nicegui.element import Element

class MonacoEditor(Element, component="monaco-editor/monaco_editor.js", exposed_libraries=["monaco-editor/lib/monaco-editor.js"]):
    
    def __init__(self, *, options: Optional[Dict] = None) -> None:
        
        super().__init__()
        # Properties in Vue properties
        self._props["options"] = options
    
    @property
    def options(self) -> Dict:
        return self._props["options"]
```

For now I got everything working, except I cannot add the ``css`` or ``ttf`` dependency files to it (for display the icons in MonacoEditor).
I have tried ``exposed_libraries=["monaco-editor/lib/*.css", "monaco-editor/lib/*.ttf"]``, but got error ``file extension doesn't supported``.

How do I add those dependency files if they are not ``.js`` extension for my custom element then?

**Comment by falkoschindler** [comment]
Hi @KiddoV,

This is a bit tricky and I'm still looking for a simpler way to integrate third-party dependencies. But you can look into `ui.leaflet` for an example:
1. Register a resource path for the Python element:
    https://github.com/zauberzeug/nicegui/blob/b7b30b2c801c6c2fe46efb0af6eef34205884163/nicegui/elements/leaflet.py#L40
2. Load the resources when mounting the Vue component:
    https://github.com/zauberzeug/nicegui/blob/b7b30b2c801c6c2fe46efb0af6eef34205884163/nicegui/elements/leaflet.js#L14-L17

**Reply by KiddoV** [reply]
Thanks, but I can only add ``CSS`` not ``TTF`` file. When I call:
```js
        await Promise.all([
            loadResource(window.path_prefix + `${this.resource_path}/monaco-editor/lib/codicon.css`),
            loadResource(window.path_prefix + `${this.resource_path}/monaco-editor/lib/codicon.ttf`)
        ]);
```
I got error:
```console
vue.global.prod.js:1 
 TypeError: Cannot read properties of undefined (reading 'setAttribute')
    at resources.js:20:17
    at new Promise (<anonymous>)
    at loadResource (resources.js:5:25)
    at Proxy.mounted (monaco_editor.js:27:13)
```

**Reply by falkoschindler** [reply]
Oh, I see. `loadResource` only knows how to handle CSS and JS files, but not fonts like TTF. I guess you don't need to call it for the TTF at all, because the font will be automatically loaded by the CSS file.

**Reply by KiddoV** [reply]
@falkoschindler I don't think so, after CSS added I don't see my TTF file anywhere in the ``DevTool Sources Tab``:
My CSS has:
```css
@font-face {
	font-family: "codicon";
	font-display: block;
	src: url("./codicon.ttf?e8353f7791e1f01b47116b67fec2f5b1") format("truetype");
}
```
The ``codicon.ttf`` file in the same folder with the CSS

**Reply by KiddoV** [reply]
I also not seeing my TTF file adding this way for some reason:
```py
app.add_static_file(url_path="/", local_file="./elements/monaco-editor/lib/codicon.ttf")
```

**Reply by falkoschindler** [reply]
The URL `url("./codicon.ttf?e8353f7791e1f01b47116b67fec2f5b1")` looks wrong. The prefix "./" is for relative file paths, but not for relative URLs. Try accessing "codicon.ttf?e8353f7791e1f01b47116b67fec2f5b1" with your browser to see if the file is correctly served.

**Reply by KiddoV** [reply]
No, I don't see the file in the browser.
This is where I get those file:
``monaco-editor.js``: https://cdn.jsdelivr.net/npm/monaco-editor@0.46.0/+esm
``codicon.css``: https://cdn.jsdelivr.net/npm/vscode-codicons@0.0.17/dist/codicon.css

In the file ``codicon.min.css`` they use ``src:url("codicon.ttf?54702f639320b60466e6f56678165405")`` instead:
https://cdn.jsdelivr.net/npm/vscode-codicons@0.0.17/dist/codicon.min.css

**Reply by falkoschindler** [reply]
Ok, maybe I was wrong about the prefix "./". Anyway, I just added the `@font-face` declaration and some `body {color: red; font-family: "codicon"}` to leaflet.css and the font is loaded successfully when using `ui.leaflet`. Not sure what is going on on your side. 😕
